### PR TITLE
docs: api docs improvement

### LINKED ISSUE #331: Improve docs and examples
- [ ] Add links to Bee's Docs and API reference
- [ ] Add `@throws` notations for the list of errors that can be thrown and in what cases
- [ ] Document gateway mode limited functions
- [ ] Document light mode limitations

### PR DESCRIPTION
**Pull request by AuHau**
Closes #331

### CODE REVIEW

File: src/bee.ts [hunk 1]
```diff
@@ -259,7 +291,12 @@ export class Bee {
   /**
    * Retrieve tag information from Bee node
    *
+   * !!! Warning! Not allowed when node is in Gateway mode!
```
Comment (nugaon) [inline src/bee.ts:294]:
```suggestion
   * **Warning! Not allowed when node is in Gateway mode!**
```

File: src/types/index.ts [hunk 2]
```diff
@@ -48,24 +56,80 @@ export type BatchId = HexString<typeof BATCH_ID_HEX_LENGTH>
...
 export interface UploadOptions {
+  /**
+   * Will pin the data locally in the Bee node as well.
+   *
+   * Locally pinned data is possible to reupload to network if it disappear.
+   *
+   * !!! Warning! Not allowed when node is in Gateway mode!
+   *
+   * @see [Bee docs - Pinning](https://docs.ethswarm.org/docs/access-the-swarm/pinning)
+   * @see [Bee API reference - `POST /bzz`](https://docs.ethswarm.org/api/#tag/File)
```
Comment (nugaon) [inline src/types/index.ts:74]: Wrong link
```suggestion
   * @see [Bee API reference - `POST /bzz`](https://docs.ethswarm.org/api/#tag/Collection/paths/~1bzz/post)
```

File: src/types/index.ts [hunk 3]
```diff
@@ -48,24 +56,80 @@ export type BatchId = HexString<typeof BATCH_ID_HEX_LENGTH>
...
+   */
   pin?: boolean
+
+  /**
+   * Will encrypt the uploaded data and return longer hash which also includes the decryption key.
+   *
+   * !!! Warning! Not allowed when node is in Gateway mode!
+   *
+   * @see [Bee docs - Store with Encryption](https://docs.ethswarm.org/docs/access-the-swarm/store-with-encryption)
+   * @see [Bee API reference - `POST /bzz`](https://docs.ethswarm.org/api/#tag/File)
```
Comment (nugaon) [inline src/types/index.ts:84]: Wrong link
```suggestion
   * @see [Bee API reference - `POST /bzz`](https://docs.ethswarm.org/api/#tag/Collection/paths/~1bzz/post)
```

File: src/types/index.ts [hunk 4]
```diff
@@ -48,24 +56,80 @@ export type BatchId = HexString<typeof BATCH_ID_HEX_LENGTH>
...
+   * @see Reference
+   */
   encrypt?: boolean
+
+  /**
+   * Tags keep track of syncing the data with network. This option allows attach existing Tag UUID to the uploaded data.
+   *
+   * @see [Bee API reference - `POST /bzz`](https://docs.ethswarm.org/api/#tag/File)
```
Comment (nugaon) [inline src/types/index.ts:92]:
```suggestion
   * @see [Bee API reference - `POST /bzz`](https://docs.ethswarm.org/api/#tag/Collection/paths/~1bzz/post)
```

File: src/types/index.ts [hunk 5]
```diff
@@ -258,7 +357,14 @@ export interface PostageBatch {
  * Options for creation of postage batch
  */
 export interface PostageBatchOptions {
+  /**
+   * Sets label for the postage batch
+   */
   label?: string
+
+  /**
+   * Sets gas price for the transaction that creates the postage batch
```
Comment (nugaon) [inline src/types/index.ts:366]:
```suggestion
   * Sets gas price in Wei for the transaction that creates the postage batch
```

File: src/utils/type.ts [hunk 6]
```diff
@@ -210,12 +210,22 @@ export function assertPublicKey(value: unknown): asserts value is PublicKey {
   assertHexString(value, PUBKEY_HEX_LENGTH, 'PublicKey')
 }
 
+/**
+ * Check if it is valid Data.
```
Comment (nugaon) [inline src/utils/type.ts:214]:
```suggestion
 * Check whether the given parameter is valid data to upload
```

File: src/utils/type.ts [hunk 7]
```diff
@@ -210,12 +210,22 @@ export function assertPublicKey(value: unknown): asserts value is PublicKey {
...
+ * @throws TypeError if not valid
+ */
 export function assertData(value: unknown): asserts value is string | Uint8Array {
   if (typeof value !== 'string' && !(value instanceof Uint8Array)) {
     throw new TypeError('Data must be either string or Uint8Array!')
   }
 }
 
+/**
+ * Check if it is a valid FileData.
```
Comment (nugaon) [inline src/utils/type.ts:225]:
```suggestion
 * Check whether the given parameter is a correct file representation to file upload.
```

File: src/bee.ts [hunk 8]
```diff
@@ -305,7 +356,12 @@ export class Bee {
   /**
    * Get pinning status of chunk with given reference
    *
+   * !!! Warning! Not allowed when node is in Gateway mode!
```
Comment (nugaon) [inline src/bee.ts:359]:
```suggestion
   * **Warning! Not allowed when node is in Gateway mode!**
```

File: src/bee.ts [hunk 9]
```diff
@@ -697,13 +809,20 @@ export class Bee {
...
    */
   async getAllPostageBatch(): Promise<PostageBatch[]> {
     return stamps.getAllPostageBatches(this.url)
   }
 
   /**
-   * Ping the base bee URL. If connection was not successful throw error
+   * Ping the base bee URL.
+   *
+   * @throws If connection was not successful throw error
```
Comment (nugaon) [inline src/bee.ts:825]: it is not really informative

**Review by nugaon — COMMENTED** [review]
I would format the warning messages with bold instead of putting exclamation marks in front of them as I mentioned below.

File: src/bee.ts [hunk 10]
```diff
@@ -226,15 +252,17 @@ export class Bee {
   }
 
   /**
-   * Upload collection of files to a Bee node
+   * Upload collection of files.
    *
-   * Uses the `fs` module of Node.js
+   * Available only in NodeJs as it uses the `fs` module.
```
Comment (agazso) [inline src/bee.ts:257]: On their website they use Node.js, I suggest we use it as well https://nodejs.org/en/
```suggestion
   * Available only in Node.js as it uses the `fs` module.
```

File: src/bee.ts [hunk 11]
```diff
@@ -542,12 +629,19 @@ export class Bee {
    * High-level function that allows you to easily set JSON data to feed.
    * JSON-like data types are supported.
    *
+   * The default Signer of Bee instance is used if `options.signer` is not specified.
+   * If non of those two is set error is thrown.
```
Comment (agazso) [inline src/bee.ts:633]:
```suggestion
   * If none of those two is set error is thrown.
```

File: src/bee.ts [hunk 12]
```diff
@@ -614,7 +710,7 @@ export class Bee {
    * Make a new feed topic from a string
    *
    * Because the topic has to be 32 bytes long this function
-   * hashes the input string to create a topic.
+   * hashes the input string to create a topic from arbitrary long string.
```
Comment (agazso) [inline src/bee.ts:713]: I'm not sure but I think you could say either `arbitrarily long string` or `string of arbitrary length`

**Review by agazso — APPROVED** [review]
Nice work!

I was wondering if we want to mention that the PSS receive and subscribe is not going to work in light mode? Strictly the API call is going to work but it will never succeed.

**Review by nugaon — APPROVED** [review]
LGTM!

**Pull request merged**